app/src/main/java/org/mozilla/reference/browser/services/AccountStateMachine.kt

```diff
@@ -0,0 +1,344 @@
+                    }
+                }
+                AccountState.AuthenticatedWithProfile -> {
+                    when (event) {
+                        Logout -> AccountState.NotAuthenticated
+                        else -> state
+                    }
+                }
+            }
+        }
+    }
+
+    private val exceptionHandler = CoroutineExceptionHandler { _, e ->
+        // TODO emit an "error" event to the state machine, and have individual states deal with it as appropriate.
+        Log.log(
+            Log.Priority.ERROR,
+            message = "Unexpected error occurred during Firefox Account state transitions. Current state: $state.",
+            throwable = e,
+            tag = logTag
+        )
+    }
+
+    private var job = Job() // TODO lazy init?
+    override val coroutineContext: CoroutineContext
+        get() = Dispatchers.IO + job + exceptionHandler
+
+    // May be accessed after Init event has been processed.
+    @Volatile private lateinit var account: FirefoxAccount
+    @Volatile private var profile: Profile? = null
+    private var state = AccountState.Start
+    private val eventQueue = ConcurrentLinkedQueue<AccountEvent>()
+
+    fun authenticatedAccount(): FirefoxAccount? {
+        return when (state) {
+            AccountState.AuthenticatedWithProfile,
+            AccountState.AuthenticatedNoProfile -> account
+            else -> null
+        }
+    }
+
+    fun accountProfile(): Profile? {
+        return when (state) {
+            AccountState.AuthenticatedWithProfile -> profile
+            else -> null
+        }
+    }
+
+    fun authenticate() {
+        event(Authenticate)
+    }
+
+    fun logout() {
+        event(Logout)
+    }
+
+    fun event(event: AccountEvent) {
+        eventQueue.add(event)
+        processQueue()
+    }
+
+    private fun processQueue() = synchronized(this) {
+        CoroutineScope(coroutineContext).launch {
+            do {
+                val event = eventQueue.poll()
+                val transitionInto = nextState(state, event)
+
+                Log.log(
+                    tag = logTag,
+                    message = "Processing event $event for state $state. Next state is $transitionInto"
+                )
+
+                state = transitionInto
+
+                stateSideEffects(state, event)?.let { successiveEvent ->
+                    Log.log(
+                        tag = logTag,
+                        message = "Ran '$event' side-effects for state $state, got successive event $successiveEvent"
+                    )
+                    eventQueue.add(successiveEvent)
+                }
+            } while (!eventQueue.isEmpty())
+        }
+    }
+
+    /**
+     * Side-effects matrix. Defines non-pure operations that must take place for state+event combinations.
+     */
+    @Suppress("ComplexMethod")
+    private suspend fun stateSideEffects(forState: AccountState, via: AccountEvent): AccountEvent? {
```

Comment:
nit: I probably would've called this `stateActions`...actions to be applied to state+event combinations, which may or may not have side effects.